pallets/token-dealer/src/lib.rs

```diff
@@ -1,155 +1,263 @@
+                .expect("Should not fail; qed");
+
+            Self::deposit_event(Event::<T>::TransferredTokensToRelayChain(who, asset_id, dest, amount));
+        }
+
+        /// Transfer `amount` of tokens to another parachain.
+        #[weight = 10]
+        pub fn transfer_assets_to_parachain_chain(
+            origin,
+            para_id: u32,
+            dest: T::AccountId,
+            dest_asset_id: Option<AssetIdOf<T>>,
+            amount: BalanceOf<T>,
+            asset_id: Option<AssetIdOf<T>>,
+        ) {
+            //TODO we don't make sure that the parachain has some tokens on the other parachain.
+            let who = ensure_signed(origin.clone())?;
+
+            let para_id: ParaId = para_id.into();
+            let para_account: T::AccountId = para_id.into_account();
+
+            if let Some(id) = asset_id {
+                let para_account = T::Lookup::unlookup(para_account.clone());
+                <assets::Module<T>>::transfer(
+                    origin, id, para_account, convert_hack(&amount)
+                )?;
+            } else {
+                T::Currency::transfer(&who, &para_account, amount, ExistenceRequirement::KeepAlive)?;
+            }
+            T::XCMPMessageSender::send_xcmp_message(
+                para_id.into(),
+                &XCMPMessage::TransferToken(dest.clone(), amount, dest_asset_id),
+            ).expect("Should not fail; qed");
+
+
+            Self::deposit_event(Event::<T>::TransferredTokensToParachain(who, asset_id, para_id, dest, dest_asset_id, amount ));
+        }
+
+        fn deposit_event() = default;
+    }
 }
 
 /// This is a hack to convert from one generic type to another where we are sure that both are the
 /// same type/use the same encoding.
 fn convert_hack<O: Decode>(input: &impl Encode) -> O {
-	input.using_encoded(|e| Decode::decode(&mut &e[..]).expect("Must be compatible; qed"))
+    input.using_encoded(|e| Decode::decode(&mut &e[..]).expect("Must be compatible; qed"))
 }
 
 impl<T: Trait> DownwardMessageHandler for Module<T> {
-	fn handle_downward_message(msg: &DownwardMessage) {
-		match msg {
-			DownwardMessage::TransferInto(dest, amount, _) => {
-				let dest = convert_hack(&dest);
-				let amount: BalanceOf<T> = convert_hack(amount);
-
-				let _ = T::Currency::deposit_creating(&dest, amount.clone());
-
-				Self::deposit_event(Event::<T>::TransferredTokensFromRelayChain(dest, amount));
-			}
-			_ => {}
-		}
-	}
+    fn handle_downward_message(msg: &DownwardMessage) {
+        match msg {
+            DownwardMessage::TransferInto(dest, relay_amount, remark) => {
+                let dest: T::AccountId = convert_hack(&dest);
+                let relay_amount: BalanceOf<T> = convert_hack(relay_amount);
+                let relay_account = RelayId::default().into_account();
+                // remark has a concerte type [u8; 32]
+                let asset_id: Option<AssetIdOf<T>> = convert_hack(remark);
```

Comment:
Are you sure, that it is possible to put asset_id in remark?

Reply:
quite certain, my understanding is that Remark is just any 32btyes data, and as long as it is scale-encoded, which we use to decode it, it is fine. Like the RMK nft uses it too.